Kudu.Core/Deployment/DeploymentManager.cs

```diff
@@ -234,10 +240,24 @@ public void Delete(string id)
                     // Perform the build deployment of this changeset
                     await Build(changeSet, tracer, deployStep, repository, deploymentInfo, deploymentAnalytics, fullBuildByDefault);
 
-                    if (!(OSDetector.IsOnWindows() && !EnvironmentHelper.IsWindowsContainers()) && _settings.RestartAppContainerOnGitDeploy())
+                    if ((!OSDetector.IsOnWindows() || EnvironmentHelper.IsWindowsContainers() || deploymentInfo.ShouldRestartAfterDeployment) && _settings.RestartAppContainerOnGitDeploy())
                     {
                         logger.Log(Resources.Log_TriggeringContainerRestart);
-                        DockerContainerRestartTrigger.RequestContainerRestart(_environment, RestartTriggerReason);
+
+                        if (_settings.RecylePreviewEnabled())
+                        {
+                            logger.Log("Triggering recycle (preview mode enabled).");
+                            tracer.Trace("Triggering recycle (preview mode enabled).");
+
+                            await RestartMainSiteAsync(deploymentInfo, logger);
+                        }
+                        else
+                        {
+                            logger.Log("Triggering recycle.");
+                            tracer.Trace("Triggering recycle.");
+
+                            DockerContainerRestartTrigger.RequestContainerRestart(_environment, RestartTriggerReason, tracer);
```

Comment:
this should go to PostDeploymentHelper .. https://github.com/projectkudu/kudu/blob/master/Kudu.Core/Helpers/PostDeploymentHelper.cs#L109-L116.

You can see that we perform many operation and this should be one of them.